test/libsolidity/SemanticTest.cpp

```diff
@@ -151,12 +162,99 @@ map<string, Builtin> SemanticTest::makeBuiltins() const
 			[this](FunctionCall const& _call) -> optional<bytes>
 			{
 				soltestAssert(_call.arguments.parameters.empty(), "No arguments expected.");
-				  return toBigEndian(u256(storageEmpty(m_contractAddress) ? 1 : 0));
-		 	}
+				return toBigEndian(u256(storageEmpty(m_contractAddress) ? 1 : 0));
+			}
+		},
+	};
+}
+
+vector<SideEffectHook> SemanticTest::makeSideEffectHooks() const
+{
+	return {
+		[](FunctionCall const& _call) -> vector<string>
+		{
+			if (_call.signature == "isoltest_side_effects_test")
+			{
+				vector<string> result;
+				for (auto const& argument: _call.arguments.parameters)
+					result.emplace_back(toHex(argument.rawBytes));
+				return result;
+			}
+			return {};
+		},
+		[this](FunctionCall const& _call) -> vector<string>
+		{
+			return eventSideEffectHook(_call);
 		}
 	};
 }
 
+vector<string> SemanticTest::eventSideEffectHook(FunctionCall const&) const
+{
+	static ABIType abiType(ABIType::Type::Hex);
+	vector<string> sideEffects;
+	vector<LogRecord> recordedLogs = ExecutionFramework::recordedLogs();
+	for (auto const& log: recordedLogs)
+	{
+		optional<EventInformation> event;
+		if (!log.topics.empty())
+			event = findEvent(log.topics[0]);
+		if (event.has_value())
+			soltestAssert(event.value().indexedTypes.size() == log.topics.size() - 1, "");
+
+		stringstream sideEffect;
+		sideEffect << "emit ";
+		if (event.has_value())
+			sideEffect << event.value().signature;
+		else
+			sideEffect << "<anonymous>";
+
+		if (m_contractAddress != log.creator)
+			sideEffect << " creator=0x" << log.creator;
+
+		vector<string> eventStrings;
+		for (auto& topic: log.topics)
+			if ((event.has_value() && topic != *log.topics.begin()) || !event.has_value())
+				eventStrings.push_back("#" + BytesUtils::formatBytes(topic.asBytes(), abiType));
```

Comment:
Not sure what is not working. I guess you think that `topics[0]` is always set, and if it's an anonymous event it will get filled by the first indexed parameter, right? If I understood correctly, `topics[0]` is only reserved for the event signature hash (if present), where indexed parameters are always placed from `topics[1]` on.

https://docs.soliditylang.org/en/v0.8.4/abi-spec.html?highlight=events#events

- `topics[0]`: `keccak(EVENT_NAME+"("+EVENT_ARGS.map(canonical_type_of).join(",")+")")` (canonical_type_of is a function that simply returns the canonical type of a given argument, e.g. for uint indexed foo, it would return uint256). If the event is declared as anonymous the topics[0] is not generated;
- `topics[n]`: `abi_encode(EVENT_INDEXED_ARGS[n - 1])` (`EVENT_INDEXED_ARGS` is the series of EVENT_ARGS that are indexed);

Reply:
Looks like that I didn't understood this part of the documentation correctly. Especially the `topics[n]: abi_encode(EVENT_INDEXED_ARGS[n - 1])` part.

Reply:
Well, looks like I was missing the fact that `findEvent()` never returns an anonymous event. In that case this part is fine. It could misrepresent an anonymous event as a non-anonymous one but we cannot really do anything about that.